File: examples/twilio-media-streams/bridge.js
'use strict';

const VOQUALIZER_HANDLER = 'plugins/a0_voqualizer/ws_voqualizer';
const INPUT_CODEC = 'pcm16/16k';
const OUTPUT_CODEC = 'pcm16/16k';
const TWILIO_SAMPLE_RATE = 8000;
const VOQUALIZER_SAMPLE_RATE = 16000;
const FRAME_HEADER_BYTES = 4;

function clamp16(value) {
  return Math.max(-32768, Math.min(32767, Math.round(value)));
}

function mulawByteToSample(byte) {
  const mu = ~byte & 0xff;
  const sign = mu & 0x80;
  const exponent = (mu >> 4) & 0x07;
  const mantissa = mu & 0x0f;
  let sample = ((mantissa << 3) + 0x84) << exponent;
  sample -= 0x84;
  return sign ? -sample : sample;
}

function sampleToMulawByte(sample) {
  const BIAS = 0x84;
  const CLIP = 32635;
  let pcm = clamp16(sample);
  let sign = 0;
  if (pcm < 0) {
    sign = 0x80;
    pcm = -pcm;
  }
  pcm = Math.min(CLIP, pcm) + BIAS;
  let exponent = 7;
  for (let mask = 0x4000; (pcm & mask) === 0 && exponent > 0; mask >>= 1) {
    exponent -= 1;
  }
  const mantissa = (pcm >> (exponent + 3)) & 0x0f;
  return (~(sign | (exponent << 4) | mantissa)) & 0xff;
}

function mulawToPcm16(mulaw) {
  const input = Buffer.from(mulaw || []);
  const out = Buffer.alloc(input.length * 2);
  for (let i = 0; i < input.length; i += 1) {
    out.writeInt16LE(clamp16(mulawByteToSample(input[i])), i * 2);
  }
  return out;
}

function pcm16ToMulaw(pcm16) {
  const input = Buffer.from(pcm16 || []);
  const samples = Math.floor(input.length / 2);
  const out = Buffer.alloc(samples);
  for (let i = 0; i < samples; i += 1) {
    out[i] = sampleToMulawByte(input.readInt16LE(i * 2));
  }
  return out;
}

function resamplePcm16Linear(pcm16, srcRate, dstRate) {
  const input = Buffer.from(pcm16 || []);
  if (srcRate === dstRate || input.length === 0) {
    return Buffer.from(input);
  }
  const inSamples = Math.floor(input.length / 2);
  const outSamples = Math.max(1, Math.round(inSamples * dstRate / srcRate));
  const out = Buffer.alloc(outSamples * 2);
  const scale = srcRate / dstRate;
  for (let i = 0; i < outSamples; i += 1) {
    const pos = i * scale;
    const left = Math.min(inSamples - 1, Math.floor(pos));
    const right = Math.min(inSamples - 1, left + 1);
    const frac = pos - left;
    const a = input.readInt16LE(left * 2);
    const b = input.readInt16LE(right * 2);
    out.writeInt16LE(clamp16(a + (b - a) * frac), i * 2);
  }
  return out;
}

function encodeVoqualizerFrame(seq, tsMs, pcm16) {
  const payload = Buffer.from(pcm16 || []);
  const frame = Buffer.alloc(FRAME_HEADER_BYTES + payload.length);
  frame.writeUInt16BE(seq & 0xffff, 0);
  frame.writeUInt16BE(tsMs & 0xffff, 2);
  payload.copy(frame, FRAME_HEADER_BYTES);
  return frame;
}

function decodeTwilioMediaPayload(payloadBase64) {
  const mulaw8k = Buffer.from(payloadBase64 || '', 'base64');
  const pcm8k = mulawToPcm16(mulaw8k);
  return resamplePcm16Linear(pcm8k, TWILIO_SAMPLE_RATE, VOQUALIZER_SAMPLE_RATE);
}

function encodeTwilioMediaPayload(pcm16k) {
  const pcm8k = resamplePcm16Linear(pcm16k, VOQUALIZER_SAMPLE_RATE, TWILIO_SAMPLE_RATE);
  return pcm16ToMulaw(pcm8k).toString('base64');
}

class TwilioVoqualizerBridge {
  constructor({ voqualizerTransport, twilioSocket, sessionId = `twilio-${Date.now().toString(36)}` } = {}) {
    this.voqualizerTransport = voqualizerTransport;
    this.twilioSocket = twilioSocket;
    this.sessionId = sessionId;
    this.bearerToken = '';
    this.streamSid = '';
    this.seq = 0;
  }

  async connect() {
    await this.voqualizerTransport.connect(VOQUALIZER_HANDLER);
    const ready = await this.voqualizerTransport.emitWithAck('voqualizer_init', {
      session_id: this.sessionId,
      asr: { codec: INPUT_CODEC },
      tts: { codec: OUTPUT_CODEC },
      barge_in: true,
    });
    if (!ready || !ready.bearer_token) {
      throw new Error('voqualizer_ready did not issue bearer_token');
    }
    this.sessionId = ready.session_id || this.sessionId;
    this.bearerToken = ready.bearer_token;
    this.voqualizerTransport.on('voqualizer_tts_chunk', (payload) => this.forwardTtsToTwilio(payload));
    this.voqualizerTransport.on('voqualizer_tts_done', () => this.sendTwilioMark('voqualizer_tts_done'));
    return ready;
  }

  async handleTwilioMessage(raw) {
    const message = typeof raw === 'string' ? JSON.parse(raw) : raw;
    if (!message || !message.event) return;
    if (message.event === 'start') {
      this.streamSid = message.start && message.start.streamSid ? message.start.streamSid : message.streamSid || '';
      return;
    }
    if (message.event === 'media') {
      await this.forwardMediaToVoqualizer(message.media || {}, message.streamSid || this.streamSid);
      return;
    }
    if (message.event === 'stop') {
      await this.voqualizerTransport.emitWithAck('voqualizer_control', this.sessionPayload({ action: 'end_session' }));
    }
  }

  async forwardMediaToVoqualizer(media, streamSid = '') {
    this.ensureBearerToken();
    const pcm16k = decodeTwilioMediaPayload(media.payload || '');
    const tsMs = Number.isFinite(Number(media.timestamp)) ? Number(media.timestamp) : this.seq * 20;
    const frame = encodeVoqualizerFrame(this.seq, tsMs, pcm16k);
    this.seq = (this.seq + 1) & 0xffff;
    await this.voqualizerTransport.emitWithAck('voqualizer_audio_chunk', this.sessionPayload({ frame, streamSid }));
  }

  forwardTtsToTwilio(payload = {}) {
    const audio = Buffer.from(payload.audio || payload.data || payload.pcm16 || []);
    const twilioPayload = encodeTwilioMediaPayload(audio);
    this.sendTwilioMedia(twilioPayload);
  }

  sendTwilioMedia(payload) {
    if (!this.twilioSocket || typeof this.twilioSocket.send !== 'function') return;
    this.twilioSocket.send(JSON.stringify({
      event: 'media',
      streamSid: this.streamSid,
      media: { payload },
    }));
  }

  sendTwilioMark(name) {
    if (!this.twilioSocket || typeof this.twilioSocket.send !== 'function') return;
    this.twilioSocket.send(JSON.stringify({
      event: 'mark',
      streamSid: this.streamSid,
      mark: { name },
    }));
  }

  sessionPayload(payload) {
    return { ...payload, bearer_token: this.bearerToken };
  }

  ensureBearerToken() {
    if (!this.bearerToken) {
      throw new Error('connect before forwarding Twilio media');
    }
  }
}

module.exports = {
  VOQUALIZER_HANDLER,
  INPUT_CODEC,
  OUTPUT_CODEC,
  TWILIO_SAMPLE_RATE,
  VOQUALIZER_SAMPLE_RATE,
  FRAME_HEADER_BYTES,
  mulawToPcm16,
  pcm16ToMulaw,
  resamplePcm16Linear,
  encodeVoqualizerFrame,
  decodeTwilioMediaPayload,
  encodeTwilioMediaPayload,
  TwilioVoqualizerBridge,
};

if (require.main === module) {
  console.log('A0 Voqualizer Twilio bridge reference module loaded. Wire ws + Socket.IO transports for production use.');
}

File: examples/asterisk-audiofork/bridge.py
"""Reference Asterisk AudioSocket bridge for a0_voqualizer.

The core conversion/forwarding logic is dependency-light and transport-injected
so it can be reviewed and tested without Asterisk, a live A0 backend, network,
credentials, or model/provider calls.
"""

from __future__ import annotations

import asyncio
import audioop
import json
import struct
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Protocol

VOQUALIZER_HANDLER = "plugins/a0_voqualizer/ws_voqualizer"
INPUT_CODEC = "pcm16/16k"
OUTPUT_CODEC = "pcm16/16k"
ASTERISK_SAMPLE_RATE = 8000
VOQUALIZER_SAMPLE_RATE = 16000
FRAME_HEADER_BYTES = 4


def resample_pcm16_linear(pcm16: bytes, src_rate: int, dst_rate: int) -> bytes:
    """Dependency-free mono PCM16 little-endian linear resampler."""

    if src_rate == dst_rate or not pcm16:
        return bytes(pcm16)
    sample_count = len(pcm16) // 2
    if sample_count == 0:
        return b""
    samples = struct.unpack("<" + "h" * sample_count, pcm16[: sample_count * 2])
    out_count = max(1, round(sample_count * dst_rate / src_rate))
    scale = src_rate / dst_rate
    out: list[int] = []
    for index in range(out_count):
        pos = index * scale
        left = min(sample_count - 1, int(pos))
        right = min(sample_count - 1, left + 1)
        frac = pos - left
        value = samples[left] + (samples[right] - samples[left]) * frac
        out.append(max(-32768, min(32767, round(value))))
    return struct.pack("<" + "h" * len(out), *out)


def asterisk_slin8_to_voqualizer_pcm16(slin8: bytes) -> bytes:
    """Convert Asterisk signed-linear 8 kHz PCM16 to Voqualizer PCM16 16 kHz."""

    return resample_pcm16_linear(slin8, ASTERISK_SAMPLE_RATE, VOQUALIZER_SAMPLE_RATE)


def voqualizer_pcm16_to_asterisk_slin8(pcm16k: bytes) -> bytes:
    """Convert Voqualizer PCM16 16 kHz to Asterisk signed-linear 8 kHz PCM16."""

    return resample_pcm16_linear(pcm16k, VOQUALIZER_SAMPLE_RATE, ASTERISK_SAMPLE_RATE)


def ulaw8_to_voqualizer_pcm16(ulaw8: bytes) -> bytes:
    """Optional G.711 μ-law 8 kHz ingress helper for non-slin Asterisk paths."""

    return asterisk_slin8_to_voqualizer_pcm16(audioop.ulaw2lin(ulaw8, 2))


def voqualizer_pcm16_to_ulaw8(pcm16k: bytes) -> bytes:
    """Optional G.711 μ-law 8 kHz egress helper for non-slin Asterisk paths."""

    return audioop.lin2ulaw(voqualizer_pcm16_to_asterisk_slin8(pcm16k), 2)


def encode_voqualizer_frame(seq: int, ts_ms: int, pcm16: bytes) -> bytes:
    """Encode A2 frame: uint16 seq + uint16 ts_ms in network byte order."""

    return struct.pack("!HH", seq & 0xFFFF, ts_ms & 0xFFFF) + bytes(pcm16)


class VoqualizerTransport(Protocol):
    async def connect(self, handler: str) -> None: ...
    async def emit_with_ack(self, event: str, payload: dict[str, Any]) -> dict[str, Any]: ...
    def on(self, event: str, callback: Callable[[dict[str, Any]], Any]) -> None: ...


class AsteriskAudioSink(Protocol):
    async def write_audio(self, pcm16_8k: bytes) -> None: ...
    async def mark(self, name: str) -> None: ...


@dataclass
class AsteriskVoqualizerBridge:
    """Transport-injected Asterisk audio-fork ↔ Voqualizer bridge."""

    voqualizer: VoqualizerTransport
    asterisk_sink: AsteriskAudioSink
    session_id: str = field(default_factory=lambda: f"asterisk-{uuid.uuid4().hex}")
    bearer_token: str = ""
    seq: int = 0
    started_at: float = field(default_factory=time.monotonic)

    async def connect(self) -> dict[str, Any]:
        await self.voqualizer.connect(VOQUALIZER_HANDLER)
        ready = await self.voqualizer.emit_with_ack(
            "voqualizer_init",
            {
                "session_id": self.session_id,
                "asr": {"codec": INPUT_CODEC},
                "tts": {"codec": OUTPUT_CODEC},
                "barge_in": True,
            },
        )
        token = ready.get("bearer_token")
        if not token:
            raise RuntimeError("voqualizer_ready did not issue bearer_token")
        self.session_id = str(ready.get("session_id") or self.session_id)
        self.bearer_token = str(token)
        self.voqualizer.on("voqualizer_tts_chunk", self.handle_tts_chunk)
        self.voqualizer.on("voqualizer_tts_done", self.handle_tts_done)
        return ready

    async def forward_asterisk_audio(self, slin8: bytes) -> dict[str, Any]:
        self.ensure_bearer_token()
        pcm16k = asterisk_slin8_to_voqualizer_pcm16(slin8)
        ts_ms = int((time.monotonic() - self.started_at) * 1000) & 0xFFFF
        frame = encode_voqualizer_frame(self.seq, ts_ms, pcm16k)
        self.seq = (self.seq + 1) & 0xFFFF
        return await self.voqualizer.emit_with_ack(
            "voqualizer_audio_chunk",
            self.session_payload({"frame": frame}),
        )

    async def end_session(self) -> dict[str, Any]:
        self.ensure_bearer_token()
        return await self.voqualizer.emit_with_ack(
            "voqualizer_control",
            self.session_payload({"action": "end_session"}),
        )

    def handle_tts_chunk(self, payload: dict[str, Any]) -> None:
        audio = payload.get("audio") or payload.get("data") or payload.get("pcm16") or b""
        pcm8k = voqualizer_pcm16_to_asterisk_slin8(bytes(audio))
        asyncio.create_task(self.asterisk_sink.write_audio(pcm8k))

    def handle_tts_done(self, payload: dict[str, Any]) -> None:
        asyncio.create_task(self.asterisk_sink.mark("voqualizer_tts_done"))

    def session_payload(self, payload: dict[str, Any]) -> dict[str, Any]:
        return {**payload, "bearer_token": self.bearer_token}

    def ensure_bearer_token(self) -> None:
        if not self.bearer_token:
            raise RuntimeError("connect before forwarding Asterisk audio")


class JsonLineAudioSocketSink:
    """Tiny sample sink for local harnesses: writes JSON-line audio records."""

    def __init__(self, writer: Any):
        self.writer = writer

    async def write_audio(self, pcm16_8k: bytes) -> None:
        self.writer.write(json.dumps({"event": "audio", "pcm16_8k_hex": pcm16_8k.hex()}).encode() + b"\n")
        await self.writer.drain()

    async def mark(self, name: str) -> None:
        self.writer.write(json.dumps({"event": "mark", "name": name}).encode() + b"\n")
        await self.writer.drain()


__all__ = [
    "VOQUALIZER_HANDLER",
    "INPUT_CODEC",
    "OUTPUT_CODEC",
    "ASTERISK_SAMPLE_RATE",
    "VOQUALIZER_SAMPLE_RATE",
    "FRAME_HEADER_BYTES",
    "resample_pcm16_linear",
    "asterisk_slin8_to_voqualizer_pcm16",
    "voqualizer_pcm16_to_asterisk_slin8",
    "ulaw8_to_voqualizer_pcm16",
    "voqualizer_pcm16_to_ulaw8",
    "encode_voqualizer_frame",
    "AsteriskVoqualizerBridge",
    "JsonLineAudioSocketSink",
]


if __name__ == "__main__":
    print("A0 Voqualizer Asterisk bridge reference module. Wire AudioSocket + Socket.IO transports for production use.")
